spyder_kernels/console/kernel.py

```diff
@@ -95,10 +95,24 @@ def __init__(self, *args, **kwargs):
         self._running_namespace = None
         self.faulthandler_handle = None
 
+        # Add handlers to control to process messages while debugging
         self.control_handlers['comm_msg'] = self.control_comm_msg
         self.control_handlers['complete_request'] = self.shell_handlers[
             'complete_request']
 
+        # Socket to signal shell_stream locally
+        self.loopback_socket = None
+
+    def record_ports(self, ports):
+        """Record ports used by shell."""
+        super().record_ports(ports)
+        # Add socket to signal shell_stream locally
+        self.shell_handlers["interrupt_eventloop"] = self._interrupt_eventloop
+        self.loopback_socket = self.shell_stream.socket.context.socket(
+            zmq.DEALER)
+        self.loopback_socket.connect(
+            "tcp://127.0.0.1:%i" % self._recorded_ports['shell_port'])
```

Comment:
I'm not so sure about this: is it really necessary to alter the Jupyter protocol like this to interrupt the eventloop? If possible, I'd really prefer to keep the current comm for this because it's simpler and it can't be broken by changes in jupyter_client, for instance.

Reply:
Well Currently you need to send a message to the frontend that will then send a message back to the kernel just to interrupt the event loop. If you have a remote kernel that is quite steep. What I do here is quite straightforward, as I am simply connecting to the shell port. If the call to _recorded_ports is the issue I could read the port from the connection file

Reply:
I tried to address your concerns in the latest commit